text/3334-niche.md

```diff
@@ -0,0 +1,480 @@
+results in an error.
+
+Declaring a niche on any item other than a struct declaration results in an
+error.
+
+Declaring a niche on a struct containing more or less than one field results in
+an error.
+
+Declaring multiple `niche` attributes on a single item, or multiple key-value
+pairs within a single `niche` attribute, results in an error.
+
+The `niche` attribute may appear inside `cfg_attr`. The net effect after
+evaluating all configuration must be to apply either zero or one `niche`
+attribute to the type.
+
+Declaring a niche on a struct that has any generic parameters results in an
+error.
+
+Declaring a range niche with an empty range (e.g. `0..0`) results in a
+warn-by-default lint. As with many lints, this lint should be automatically
+suppressed for code expanded from a macro.
+
+Declaring a range niche with an invalid range (e.g. `5..0`) results in an
+error.
+
+Declaring a niche using a negative value or a negative range endpoint results
+in an error. The representation of negative values depends on the size of the
+type, and the compiler may not have that information at the time it handles
+attributes such as `niche`. The text of the error should suggest the
+appropriate two's-complement unsigned equivalent to use. The compiler may
+support this in the future.
+
+Declaring a range niche with an open start (`..3`) results in an error, for
+forwards-compatibility with support for negative values.
+
+Declaring a niche using a non-literal value (e.g. `usize::MAX`) results in an
+error. Constants can use compile-time evaluation, and compile-time evaluation
+does not occur early enough for attributes such as niche declarations.
+
+If a type `T` contains multiple niche values (e.g. `#[niche(range = 8..16)]`),
+the compiler does not guarantee any particular usage of those niche values in
+the representation of types containing `T`. In particular, the compiler does
+not commit to making use of all the niche values, even if it otherwise could
+have.
+
+However, multiple instances of the same identical type (e.g. `Option<T>` and
+`Option<T>`) will use an identical representation (whether the type contains a
+niche or not). This permits a round-trip between such a value and a byte
+representation.
+
+Adding a niche to a type does not change the storage size of the type, even if
+the niche might otherwise allow storing fewer bytes. The type still allows
+obtaining mutable references to the field, which requires storing valid values
+using the same representation as those values would have had without the niche.
+
+Declaring a niche *may* allow additional optimizations that assume the type
+cannot contain the niche values, though the compiler does not guarantee this.
+For instance, the compiler may be able to elide bounds checks that the valid
+values always satisfy.
+
+Niches do not affect pattern-matching exhaustiveness. For the purposes of
+pattern matching, the compiler will check exhaustiveness as if the field could
+take on any value.
```

Comment:
I would comment at least a bit on the future compatibility of this, since it's worth understanding how changes in the future might affect this.

For example, would annotating a niche struct with `#[non_exhaustive]` effectively force all consumers of the type to ignore the niche and specify the remaining pattern, or would this change be applied across library boundaries regardless? Even though all callers currently will not be able to assume exhaustiveness, I think that specifying where this might potentially affect API stability, and what you can do to prevent it, would be a good idea.

Personally, I like the `#[non_exhaustive]` approach, but could be persuaded to another approach.